File: python/src/tdcv2/format/transforms.py
"""Case, slicing, grouping, escaping — the filters that shape a value on its way out.

Pure text in, text out, shared by three call sites that all mean the same thing: compute tags
(``<upper>``, ``<mask>``), the ``case=``/``mask=`` attributes of a ``<gen>``, and the
``${{Name|upper|mask:xxx}}`` filters inside a ``<data>``. One implementation, so a value formatted
in a template and the same value formatted by an attribute cannot drift apart.

The filters are lenient on purpose. An out-of-range slice yields less text, a ``compact`` of
something that is not a number passes it through — the validator is where a mistake gets named,
and aborting a million-row run over one short value is worse than a gap.
"""

from __future__ import annotations

import re
from collections.abc import Callable

from ..lib import numbers, text
from .mask import apply_mask

_WHOLE = re.compile(r"^-?\d+$")

CASE_TRANSFORMS = ("upper", "lower", "capitalize", "title")

FILTER_NAMES = (
    "upper",
    "lower",
    "capitalize",
    "title",
    "mask",
    "slice",
    "replace",
    "trim",
    "group",
    "compact",
    "csv",
    "sql",
)


def is_case_transform(name: str) -> bool:
    return name in CASE_TRANSFORMS


def is_filter_name(name: str) -> bool:
    return name in FILTER_NAMES


def gen_formatter(mask: str | None, case: str | None) -> Callable[[str], str] | None:
    """The per-value formatter for a gen's ``mask=``/``case=``, or ``None`` when neither applies.

    Mask first, then case: the mask decides which characters survive and in what order, and
    changing the case of characters that are about to be dropped is wasted work.
    """
    has_mask = mask is not None
    has_case = case is not None and is_case_transform(case)
    if not has_mask and not has_case:
        return None

    def format_one(value: str) -> str:
        out = value
        if has_mask:
            out = apply_mask(mask, out)  # type: ignore[arg-type]
        if has_case:
            out = apply_case(case, out)  # type: ignore[arg-type]
        return out

    return format_one


def apply_case(name: str, value: str) -> str:
    """``capitalize`` and ``title`` raise the first letter and leave the rest as written.

    Lowering the rest would turn ``McDonald`` into ``Mcdonald`` and ``IBM`` into ``Ibm``. A name
    is not a sentence, and the config already chose how the value was spelled.
    """
    if name == "upper":
        return value.upper()
    if name == "lower":
        return value.lower()
    if name == "capitalize":
        return _upper_first(value)
    if name == "title":
        return text.NOT_SPACE_RUN.sub(lambda m: _upper_first(m.group()), value)
    raise ValueError(f'unknown case transform "{name}"')


def apply_slice(value: str, start: float, end: float | None = None) -> str:
    """A substring by character index, ``[start, end)``; an absent end runs to the end."""
    chars = list(value)
    first = int(start) if _finite(start) else 0
    last = len(chars) if end is None or not _finite(end) else int(end)
    return "".join(chars[first:last])


def apply_replace(value: str, old: str, new: str) -> str:
    return value if old == "" else value.replace(old, new)


def apply_trim(value: str) -> str:
    return text.trim(value)


def apply_group(value: str, size: float, separator: str) -> str:
    """Characters grouped from the RIGHT into chunks of ``size``.

    From the right because that is where the significant end of a number is: grouping
    ``1234567`` from the left gives ``123 456 7``, which no one writes.
    """
    chars = list(value)
    if not _finite(size) or size <= 0 or not chars:
        return value
    step = int(size)
    out: list[str] = []
    end = len(chars)
    while end > 0:
        out.insert(0, "".join(chars[max(0, end - step) : end]))
        end -= step
    return separator.join(out)


def apply_compact(value: str, base: float) -> str:
    """A whole number rendered in a shorter alphabet — ``1000000`` becomes ``lfls``.

    The point is a unique suffix that stays readable at scale: a row id appended to a generated
    email keeps it unique, but ``john.smith2000000000@`` is nobody's address. In base 36 the same
    id is ``x2qxvk``, and six characters carry over two billion rows.

    LOWERCASE only, and deliberately so. Base 62 would be shorter still, but many systems fold the
    local part of an address to lower case — ``aB`` and ``Ab`` would merge and silently
    reintroduce exactly the duplicates the suffix exists to prevent.
    """
    body = value.strip()
    if not _WHOLE.match(body):
        return value
    if base != int(base) or base < 2 or base > 36:
        return value
    n = int(body)
    if abs(n) > 9007199254740991:
        return value
    return ("-" if n < 0 else "") + _to_base(abs(n), int(base))


def apply_csv(value: str, _delimiter: str) -> str:
    """A value quoted for CSV per RFC 4180: wrapped in quotes, with any inside doubled.

    ``<data>`` assembles TEXT and knows nothing about the file being written, so a value carrying
    the delimiter silently splits the row. Measured on a real catalogue: one product named
    ``Набор ножей, 3 шт`` turned 6172 of 50 000 rows into eight fields where the header declares
    seven — category landing in price, price in quantity — with no error and nothing visibly wrong.

    Quoting unconditionally rather than only when needed: a rule with no exceptions is one nobody
    has to remember, every CSV reader accepts redundant quotes, and "only when it contains a
    comma" is exactly the reasoning that loses to a newline or a quote later.
    """
    doubled = value.replace('"', '""')
    return f'"{doubled}"'


def apply_sql(value: str) -> str:
    """A value escaped for a single-quoted SQL literal, by doubling apostrophes.

    ``O'Brien`` closes the string early and the statement fails to parse — or, far worse in
    generated data, parses into something else. The BODY only, no surrounding quotes, so the
    caller keeps writing ``'${{Name|sql}}'`` and the shape of the statement stays visible in the
    config.
    """
    return value.replace("'", "''")


def apply_filter(kind: str, arg: str | None, value: str) -> str:
    """One filter applied to one value; an unknown name is a no-op the validator flags.

    Argument shapes: ``mask:PATTERN``, ``slice:from[,to]``, ``replace:from,to``, ``trim``,
    ``group:size[,sep]`` (a space by default), ``compact[:base]``, ``csv[:delimiter]``, ``sql``,
    and the four case transforms, which take none.
    """
    if kind == "mask":
        return apply_mask(arg or "", value)
    if kind == "slice":
        parts = (arg or "").split(",")
        start = numbers.parse(parts[0] if parts else "0")
        end = None if len(parts) < 2 or parts[1] == "" else numbers.parse(parts[1])
        return apply_slice(value, start, end)
    if kind == "replace":
        old, new = _split_first(arg or "", ",")
        return apply_replace(value, old, new or "")
    if kind == "trim":
        return apply_trim(value)
    if kind == "group":
        size, separator = _split_first(arg or "", ",")
        joiner = " " if separator is None else separator
        return apply_group(value, numbers.parse(size or "3"), joiner)
    if kind == "compact":
        return apply_compact(value, 36 if not arg else numbers.parse(arg))
    if kind == "csv":
        return apply_csv(value, arg if arg else ",")
    if kind == "sql":
        return apply_sql(value)
    if kind in CASE_TRANSFORMS:
        return apply_case(kind, value)
    return value


def _upper_first(word: str) -> str:
    return word[:1].upper() + word[1:] if word else word


def _split_first(value: str, separator: str) -> tuple[str, str | None]:
    i = value.find(separator)
    return (value, None) if i < 0 else (value[:i], value[i + 1 :])


def _finite(value: float) -> bool:
    return value == value and value not in (float("inf"), float("-inf"))


def _to_base(n: int, base: int) -> str:
    if n == 0:
        return "0"
    digits = "0123456789abcdefghijklmnopqrstuvwxyz"
    out: list[str] = []
    while n:
        n, rest = divmod(n, base)
        out.append(digits[rest])
    return "".join(reversed(out))

File: csharp/Tdcv2/Format/Transforms.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Tdcv2.Format;

/// <summary>
/// Text transforms applied to a finished value.
/// </summary>
/// <remarks>
/// Shared by three places that all mean the same thing: the <c>case=</c> attribute on a
/// <c>&lt;gen&gt;</c>, the compute tags, and the <c>${{Name|upper}}</c> interpolation filters. One
/// implementation, so the three cannot drift apart.
/// </remarks>
public static class Transforms
{
    public static IReadOnlyList<string> CaseTransforms { get; } =
        new[] { "upper", "lower", "capitalize", "title" };

    /// <summary>Every name accepted after a <c>|</c> inside an interpolation.</summary>
    public static IReadOnlyList<string> FilterNames { get; } = new[]
    {
        "upper", "lower", "capitalize", "title", "mask", "slice", "replace", "trim", "group",
        "compact", "csv", "sql",
    };

    /// <summary>A run of non-whitespace, for <c>title</c>.</summary>
    private static readonly Regex Word = new(@"\S+", RegexOptions.Compiled);

    private static readonly Regex WholeNumber = new(@"^-?\d+$", RegexOptions.Compiled);

    public static bool IsCaseTransform(string name) => CaseTransforms.Contains(name);

    public static bool IsFilterName(string name) => FilterNames.Contains(name);

    /// <summary>
    /// Apply one interpolation filter.
    /// </summary>
    /// <remarks>
    /// An unknown filter passes the value through untouched. Filters are lenient by design and the
    /// validator is where a typo gets named; failing here would turn a misspelling into a dead run
    /// rather than a visible oddity in the output.
    /// </remarks>
    public static string ApplyFilter(string kind, string? arg, string value)
    {
        string a = arg ?? "";
        switch (kind)
        {
            case "mask":
                return Mask.Apply(a, value);
            case "slice":
            {
                string[] parts = a.Split(',');
                int? to = parts.Length < 2 || parts[1].Length == 0 ? null : IntOr(parts[1], null);
                return Slice(value, parts.Length > 0 ? IntOr(parts[0], 0) ?? 0 : 0, to);
            }

            case "replace":
            {
                int comma = a.IndexOf(',');
                string from = comma < 0 ? a : a[..comma];
                string to = comma < 0 ? "" : a[(comma + 1)..];
                return from.Length == 0 ? value : value.Replace(from, to);
            }

            case "trim":
                return value.Trim();
            case "group":
            {
                int comma = a.IndexOf(',');
                string size = comma < 0 ? a : a[..comma];
                string sep = comma < 0 ? " " : a[(comma + 1)..];
                return Group(value, size.Length == 0 ? 3 : IntOr(size, 3) ?? 3, sep);
            }

            case "compact":
                return Compact(value, a.Length == 0 ? 36 : IntOr(a, 36) ?? 36);
            case "csv":
                return Csv(value);
            case "sql":
                return Sql(value);
            case "upper":
            case "lower":
            case "capitalize":
            case "title":
                return ApplyCase(kind, value);
            default:
                return value;
        }
    }

    /// <summary>
    /// A substring by code-point index, <c>[from, to)</c>; a missing <c>to</c> means the end.
    /// </summary>
    /// <remarks>
    /// A negative index counts from the END, which is what the reference's <c>Array.slice</c> does
    /// and what Python's own slicing does. It matters: <c>slice:-3</c> has to mean "the last three
    /// characters" everywhere, not "all of them" in whichever implementation clamped it to zero.
    /// </remarks>
    public static string Slice(string s, int from, int? to)
    {
        IReadOnlyList<string> cp = Mask.CodePoints(s);
        int n = cp.Count;
        int f = from < 0 ? Math.Max(0, n + from) : Math.Min(from, n);
        int t = to is null ? n : (to.Value < 0 ? Math.Max(0, n + to.Value) : Math.Min(to.Value, n));
        return t <= f ? "" : string.Concat(cp.Skip(f).Take(t - f));
    }

    /// <summary>Group characters from the <em>right</em>, so a number's last group stays whole.</summary>
    public static string Group(string s, int size, string sep)
    {
        if (size <= 0 || s.Length == 0)
        {
            return s;
        }

        // A decimal number is grouped where a person groups one: the digits BEFORE the
        // separator, and nowhere else. Chunking the whole string from the right put the space in
        // the fraction — 1234.56 came out "1 234 .56", a number in no locale, and nothing said
        // so. Only this exact shape is treated as a number, so group:4 on a card number stays the
        // blocks it was written for, and so does every other string.
        System.Text.RegularExpressions.Match decimalValue = DecimalShape.Match(s);
        if (decimalValue.Success)
        {
            return decimalValue.Groups[1].Value
                + ChunkFromRight(decimalValue.Groups[2].Value, size, sep)
                + decimalValue.Groups[3].Value;
        }

        return ChunkFromRight(s, size, sep);
    }

    private static readonly System.Text.RegularExpressions.Regex DecimalShape =
        new(@"^([+-]?)(\d+)(\.\d+)$");

    private static string ChunkFromRight(string s, int size, string sep)
    {
        IReadOnlyList<string> cp = Mask.CodePoints(s);
        if (cp.Count == 0)
        {
            return s;
        }

        var parts = new List<string>();
        for (int end = cp.Count; end > 0; end -= size)
        {
            int start = Math.Max(0, end - size);
            parts.Insert(0, string.Concat(cp.Skip(start).Take(end - start)));
        }

        return string.Join(sep, parts);
    }

    /// <summary>
    /// Write a whole number in a shorter alphabet: <c>1000000</c> becomes <c>lfls</c>.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The point is a unique suffix that stays readable at scale. A row id appended to a generated
    /// address keeps it unique, but <c>john.smith2000000000@</c> is nobody's email; in base 36 the
    /// same id is six characters and covers two billion rows.
    /// </para>
    /// <para>
    /// Lowercase only, and deliberately. Base 62 would be shorter, but many systems fold the local
    /// part of an address to lower case, so <c>aB</c> and <c>Ab</c> would merge and quietly
    /// reintroduce the duplicates the suffix exists to prevent.
    /// </para>
    /// </remarks>
    public static string Compact(string value, int radix)
    {
        string text = value.Trim();
        if (!WholeNumber.IsMatch(text) || radix < 2 || radix > 36)
        {
            return value;
        }

        if (!long.TryParse(text, NumberStyles.Integer, CultureInfo.InvariantCulture, out long n))
        {
            return value;
        }

        return (n < 0 ? "-" : "") + ToRadix(Math.Abs(n), radix);
    }

    /// <summary>
    /// Quote a value for CSV, per RFC 4180.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <c>&lt;data&gt;</c> assembles text and knows nothing about the file being written, so a
    /// value containing the delimiter silently splits the row — a product named
    /// <c>Knife set, 3 pcs</c> turns one seven-field row into eight fields, with category landing
    /// in price and price in quantity, and nothing anywhere reporting an error.
    /// </para>
    /// <para>
    /// Quoted unconditionally rather than only when needed: a rule with no exceptions is one nobody
    /// has to remember, every reader accepts redundant quotes, and "only when it contains a comma"
    /// is exactly the reasoning that loses to a newline later.
    /// </para>
    /// </remarks>
    public static string Csv(string value) => "\"" + value.Replace("\"", "\"\"") + "\"";

    /// <summary>
    /// Escape a value for a single-quoted SQL literal by doubling apostrophes.
    /// </summary>
    /// <remarks>
    /// <c>O'Brien</c> closes the string early and the statement fails to parse — or worse, in
    /// generated data, parses into something else. The body only, with no surrounding quotes, so
    /// the config keeps writing <c>'${{Name|sql}}'</c> and the shape of the statement stays visible
    /// where it is written.
    /// </remarks>
    public static string Sql(string value) => value.Replace("'", "''");

    public static string ApplyCase(string name, string s) => name switch
    {
        "upper" => Upper(s),
        "lower" => s.ToLowerInvariant(),
        "capitalize" => UpperFirst(s),
        // Only the first letter of each word moves; the rest is left as written, so an
        // already-correct "McDonald" is not flattened to "Mcdonald".
        "title" => TitleCase(s),
        _ => throw new ArgumentException($"unknown case transform \"{name}\""),
    };

    /// <summary>
    /// Uppercase with FULL case mapping, the way JavaScript, Java and Python all do it.
    /// </summary>
    /// <remarks>
    /// .NET is the odd one out here: <c>ToUpperInvariant</c> maps one character to one character,
    /// so German <c>ß</c> stays <c>ß</c> where every other implementation writes <c>SS</c>. One
    /// row in a hundred thousand would differ, which is exactly the kind of divergence that gets
    /// found by a customer rather than by a test. The table is the set of code points whose
    /// uppercase is longer than one character, taken from the reference itself rather than
    /// guessed — regenerate it the same way if Unicode grows another.
    /// </remarks>
    public static string Upper(string s)
    {
        var result = new StringBuilder(s.Length);
        for (int i = 0; i < s.Length; i++)
        {
            int cp = char.ConvertToUtf32(s, i);
            if (char.IsHighSurrogate(s[i]))
            {
                i++;
            }

            if (MultiCharUpper.TryGetValue(cp, out string? expanded))
            {
                result.Append(expanded);
            }
            else
            {
                result.Append(char.ConvertFromUtf32(cp).ToUpperInvariant());
            }
        }

        return result.ToString();
    }

    /// <summary>Code points whose uppercase is more than one character.</summary>
    private static readonly IReadOnlyDictionary<int, string> MultiCharUpper =
        new Dictionary<int, string>
        {
        [0x00DF] = "SS",
        [0x0149] = "\u02bcN",
        [0x01F0] = "J\u030c",
        [0x0390] = "\u0399\u0308\u0301",
        [0x03B0] = "\u03a5\u0308\u0301",
        [0x0587] = "\u0535\u0552",
        [0x1E96] = "H\u0331",
        [0x1E97] = "T\u0308",
        [0x1E98] = "W\u030a",
        [0x1E99] = "Y\u030a",
        [0x1E9A] = "A\u02be",
        [0x1F50] = "\u03a5\u0313",
        [0x1F52] = "\u03a5\u0313\u0300",
        [0x1F54] = "\u03a5\u0313\u0301",
        [0x1F56] = "\u03a5\u0313\u0342",
        [0x1F80] = "\u1f08\u0399",
        [0x1F81] = "\u1f09\u0399",
        [0x1F82] = "\u1f0a\u0399",
        [0x1F83] = "\u1f0b\u0399",
        [0x1F84] = "\u1f0c\u0399",
        [0x1F85] = "\u1f0d\u0399",
        [0x1F86] = "\u1f0e\u0399",
        [0x1F87] = "\u1f0f\u0399",
        [0x1F88] = "\u1f08\u0399",
        [0x1F89] = "\u1f09\u0399",
        [0x1F8A] = "\u1f0a\u0399",
        [0x1F8B] = "\u1f0b\u0399",
        [0x1F8C] = "\u1f0c\u0399",
        [0x1F8D] = "\u1f0d\u0399",
        [0x1F8E] = "\u1f0e\u0399",
        [0x1F8F] = "\u1f0f\u0399",
        [0x1F90] = "\u1f28\u0399",
        [0x1F91] = "\u1f29\u0399",
        [0x1F92] = "\u1f2a\u0399",
        [0x1F93] = "\u1f2b\u0399",
        [0x1F94] = "\u1f2c\u0399",
        [0x1F95] = "\u1f2d\u0399",
        [0x1F96] = "\u1f2e\u0399",
        [0x1F97] = "\u1f2f\u0399",
        [0x1F98] = "\u1f28\u0399",
        [0x1F99] = "\u1f29\u0399",
        [0x1F9A] = "\u1f2a\u0399",
        [0x1F9B] = "\u1f2b\u0399",
        [0x1F9C] = "\u1f2c\u0399",
        [0x1F9D] = "\u1f2d\u0399",
        [0x1F9E] = "\u1f2e\u0399",
        [0x1F9F] = "\u1f2f\u0399",
        [0x1FA0] = "\u1f68\u0399",
        [0x1FA1] = "\u1f69\u0399",
        [0x1FA2] = "\u1f6a\u0399",
        [0x1FA3] = "\u1f6b\u0399",
        [0x1FA4] = "\u1f6c\u0399",
        [0x1FA5] = "\u1f6d\u0399",
        [0x1FA6] = "\u1f6e\u0399",
        [0x1FA7] = "\u1f6f\u0399",
        [0x1FA8] = "\u1f68\u0399",
        [0x1FA9] = "\u1f69\u0399",
        [0x1FAA] = "\u1f6a\u0399",
        [0x1FAB] = "\u1f6b\u0399",
        [0x1FAC] = "\u1f6c\u0399",
        [0x1FAD] = "\u1f6d\u0399",
        [0x1FAE] = "\u1f6e\u0399",
        [0x1FAF] = "\u1f6f\u0399",
        [0x1FB2] = "\u1fba\u0399",
        [0x1FB3] = "\u0391\u0399",
        [0x1FB4] = "\u0386\u0399",
        [0x1FB6] = "\u0391\u0342",
        [0x1FB7] = "\u0391\u0342\u0399",
        [0x1FBC] = "\u0391\u0399",
        [0x1FC2] = "\u1fca\u0399",
        [0x1FC3] = "\u0397\u0399",
        [0x1FC4] = "\u0389\u0399",
        [0x1FC6] = "\u0397\u0342",
        [0x1FC7] = "\u0397\u0342\u0399",
        [0x1FCC] = "\u0397\u0399",
        [0x1FD2] = "\u0399\u0308\u0300",
        [0x1FD3] = "\u0399\u0308\u0301",
        [0x1FD6] = "\u0399\u0342",
        [0x1FD7] = "\u0399\u0308\u0342",
        [0x1FE2] = "\u03a5\u0308\u0300",
        [0x1FE3] = "\u03a5\u0308\u0301",
        [0x1FE4] = "\u03a1\u0313",
        [0x1FE6] = "\u03a5\u0342",
        [0x1FE7] = "\u03a5\u0308\u0342",
        [0x1FF2] = "\u1ffa\u0399",
        [0x1FF3] = "\u03a9\u0399",
        [0x1FF4] = "\u038f\u0399",
        [0x1FF6] = "\u03a9\u0342",
        [0x1FF7] = "\u03a9\u0342\u0399",
        [0x1FFC] = "\u03a9\u0399",
        [0xFB00] = "FF",
        [0xFB01] = "FI",
        [0xFB02] = "FL",
        [0xFB03] = "FFI",
        [0xFB04] = "FFL",
        [0xFB05] = "ST",
        [0xFB06] = "ST",
        [0xFB13] = "\u0544\u0546",
        [0xFB14] = "\u0544\u0535",
        [0xFB15] = "\u0544\u053b",
        [0xFB16] = "\u054e\u0546",
        [0xFB17] = "\u0544\u053d"
        };

    private static int? IntOr(string raw, int? fallback) =>
        int.TryParse(raw.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out int value)
            ? value
            : fallback;

    /// <summary>Java's <c>Long.toString(n, radix)</c>: lowercase digits, no padding.</summary>
    private static string ToRadix(long value, int radix)
    {
        if (value == 0)
        {
            return "0";
        }

        const string digits = "0123456789abcdefghijklmnopqrstuvwxyz";
        var buffer = new StringBuilder();
        while (value > 0)
        {
            buffer.Insert(0, digits[(int)(value % radix)]);
            value /= radix;
        }

        return buffer.ToString();
    }

    private static string TitleCase(string s)
    {
        var result = new StringBuilder();
        int last = 0;
        foreach (Match m in Word.Matches(s))
        {
            result.Append(s, last, m.Index - last).Append(UpperFirst(m.Value));
            last = m.Index + m.Length;
        }

        return result.Append(s[last..]).ToString();
    }

    /// <summary>Uppercase the first character by code point, so a surrogate pair is not split.</summary>
    private static string UpperFirst(string word)
    {
        if (word.Length == 0)
        {
            return word;
        }

        int width = char.IsHighSurrogate(word[0]) && word.Length > 1 ? 2 : 1;
        return word[..width].ToUpperInvariant() + word[width..];
    }
}
